Bazel for embedded

blaizard opened the pull request:
This is an article proposal about presenting the build system Bazel in the context of embedded development.

bahildebrand commented on _drafts/bazel-for-embedded.md:
```diff
@@ -0,0 +1,383 @@
+
+This is it! You have integrated a semi-hermetic toolchain into Bazel!
+
+Note, `--incompatible_enable_cc_toolchain_resolution` is required to enable toolchain resolution with platforms for C++ toolchains. This flag should be flipped in future Bazel releases and therefore not be needed anymore.
```
Not sure what `flipped` means in this context. Seems like we can just say it's no longer needed?

blaizard replied:
This is some kind of Bazel jargon that usually people use. Yes it means that it will not be needed anymore, I will clarify that.

bahildebrand commented on _drafts/bazel-for-embedded.md:
```diff
@@ -0,0 +1,383 @@
+
+Note, `--incompatible_enable_cc_toolchain_resolution` is required to enable toolchain resolution with platforms for C++ toolchains. This flag should be flipped in future Bazel releases and therefore not be needed anymore.
+
+The code described in this section can be found here: https://github.com/blaizard/bazel-examples/tree/main/toolchain-arm64-gcc
```
Might want to just make this a markdown link. It just renders as text in the preview.

bahildebrand commented on _drafts/bazel-for-embedded.md:
```diff
@@ -0,0 +1,383 @@
+
+The last step is to tell Bazel to prioritize Docker sandboxing over the default strategy. That way, Bazel will check if the execution platform selected is compatible with the Docker strategy and use it, if not, the default sandboxing will be used. 
+
+This is done by adding these 2 lines in the `.bazelrc` file:
```
This is the first mention of the `.bazelrc` so far. Maybe it would be worthwhile to have a few sentences on what it is, and where it lives.

blaizard replied:
Good point, I will add a short section at the beginning to introduce some concepts.

bahildebrand commented on _drafts/bazel-for-embedded.md:
```diff
@@ -0,0 +1,383 @@
+
+The code described in this section can be found here: https://github.com/blaizard/bazel-examples/tree/main/toolchain-arm64-gcc
+
+### Dockerized
```
This is awesome!

bahildebrand commented on _drafts/bazel-for-embedded.md:
```diff
@@ -0,0 +1,383 @@
+
+### Dockerized
+
+Often when dealing with cross compilation, you don’t get a self contained binary or package that you can just unzip and use directly, but you need to install the toolchain on your machine, which makes it very difficult to use or integrate in CI for example or might conflict with an existing installation you already have.
```
Might be good to touch on how this can make it easier for developers to set up their environment as well.

bahildebrand reviewed (commented):
Really great draft! I had a few comments, but overall it looks good!

There are some grammatical fixups that are needed, but usually we resolve these by running it through grammarly. If you wouldn't mind doing that before your next rev it would be greatly appreciated!

blaizard commented on _drafts/bazel-for-embedded.md:
```diff
@@ -0,0 +1,383 @@
+- Using different compilers.
+- Sharing software components.
+- Testing on different architecture.
+- ...
```
No just examples on where build variants can be useful.
I reformatted it that might help.

noahp commented on _drafts/bazel-for-embedded.md:
```diff
@@ -0,0 +1,389 @@
+
+However, Bazel has also a steep learning curve, which may pose a challenge for beginners. The effort put into mastering its main concepts is worthwhile, as it offers numerous advantages over traditional build systems such as Makefile and CMake.
+
+By leveraging Bazel's capabilities, developers can achieve faster build times, better reproducibility, and higher levels of code optimization, ultimately leading to more efficient and reliable embedded systems. Therefore, it is worth investing the time and resources necessary to gain a deep understanding of Bazel and incorporate it into your development workflow.
```
Can you add a set of References for further reading? Would be helpful, especially if you know of some excellent public examples of `.bazelrc` files used for embedded.

blaizard replied:
There are not much many bazel/embedded related resources out-there, at least as I know of but I will add some useful links for further reading.

noahp commented on _drafts/bazel-for-embedded.md:
```diff
@@ -0,0 +1,389 @@
+
+This process is transparent to the user and therefore, if multiple toolchains are available for various execution platforms (for example, one for Linux, one for Windows), the user can seamlessly compile the same code on different hosts running the same command without even thinking about it.
+
+## Toolchains
```
Could you add a section before `## Toolchains` that does the following:

1. intro to executing `bazel` from the command line (just a few examples to set the context, and link to a full tutorial), explanation of a couple of the basic commands (`bazel run` eg)
2. step-by-step explained `.bazelrc` file (eg as in the below `cc_binary()` snippet) that builds a basic hello world c/cpp program, to provide context

Otherwise we're jumping right into toolchain management without setting some background on basic bazel project setup

blaizard replied:
Yes make sense, I will add the section before "Platforms" even, so just after introducing "Bazel".

blaizard commented on _drafts/bazel-for-embedded.md:
```diff
@@ -0,0 +1,389 @@
+---
+title: "Bazel for embedded"
```
I am tending toward: "Bazel Build System for Embedded Projects"

blaizard commented:
Thanks @noahp for the review, I just addressed your comments with https://github.com/memfault/interrupt/pull/319/commits/2f21a57caf6cae8f4a7ceedd62d73488398c30eb

bahildebrand commented:
Hey @blaizard, I think we're in a good place to publish this. Can you move your post from `_drafts` to `_posts` and prepend it with tomorrow's date (`2023-05-31`). You can look at previous posts to verify the naming format.

bahildebrand commented:
> Hey @blaizard, I think we're in a good place to publish this. Can you move your post from `_drafts` to `_posts` and prepend it with tomorrow's date (`2023-05-31`). You can look at previous posts to verify the naming format.

Actually just going to do this in a follow up PR. Thanks for the post!
